### PR TITLE
container/store: Process whiteouts

### PR DESCRIPTION
**Pull request by cgwalters**
A long time ago, in a galaxy far far away, a few of us in the
ostree space did some initial work on Docker-related things; specifically
support for whiteouts landed in
https://github.com/ostreedev/ostree/commit/baaf7450da8a3870e8a42f0cdd4e0ea0ed5018d6

I wish at that time we'd realized how we could more natively
support fetching containers; but, it never occurred to me to fork
off skopeo to do all the heavy lifting for the *fetch* side, which
would have been a lot of work to reimplement particularly in C.
Oh well, better late than never!

Anyways, that whiteout processing was only designed to happen at
checkout time - i.e. when materializing the final filesystem tree.  I
think this was actually a misdesign and we should add
`ostree_mutable_tree_write_with_whiteouts` so that the whiteouts
are processed in-memory.

However for now, there's a relatively low cost to temporarily
materializing the merged tree via hardlinks and handle whiteouts
via the existing code, so let's do that.

Closes: https://github.com/ostreedev/ostree-rs-ext/issues/273

---

### CODE REVIEW

**Comment by cgwalters** [comment]
Draft since this passes unit test CI, but I want to test it out for real on a host system and more carefully check that e.g. we're not doing anything stupid like physical file copies during the checkout/commit phase.

**Comment by cgwalters** [comment]
OK yep, been playing with this some more and it looks like it's working quite well!

File: lib/src/ostree_manual.rs [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+pub fn repo_file_read(f: &ostree::RepoFile) -> Result<gio::InputStream, glib::Error> {
+    use glib::translate::*;
+    let stream = unsafe {
+        let f = f.upcast_ref::<gio::File>();
+        let mut error = ptr::null_mut();
+        let stream = gio::ffi::g_file_read(f.to_glib_none().0, ptr::null_mut(), &mut error);
+        if !error.is_null() {
+            return Err(from_glib_full(error));
+        }
+        let stream = stream as *mut gio::ffi::GInputStream;
```
Comment (lucab) [inline lib/src/ostree_manual.rs:21]: It would be good to make sure that `stream` is non-null before doing this.

File: lib/src/ostree_manual.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+use std::io::Read;
+use std::ptr;
+
+use ostree;
+use ostree::prelude::{Cast, InputStreamExtManual};
+use ostree::{gio, glib};
+
+#[allow(unsafe_code)]
+
+/// Equivalent of `g_file_read()` for ostree::RepoFile to work around an ostree bug.
```
Comment (lucab) [inline lib/src/ostree_manual.rs:10]: Let's explicitly link https://github.com/ostreedev/ostree/issues/2703 here maybe?

**Review by lucab — APPROVED** [review]

File: lib/src/ostree_manual.rs (hunk 1 above)
Comment (cgwalters) [reply lib/src/ostree_manual.rs:21]: We're just casting the pointer, not dereferencing here.  The dereference happens in `from_glib_full`, which already has a null check: https://docs.rs/glib/0.14.4/src/glib/shared.rs.html#422

**Comment by cgwalters** [comment]
Also dropped the manual `g_file_query_info()` wrapper - I somehow confused myself into thinking that was necessary, but it isn't AFAICS.

**Pull request merged**